src/cookies.rs

```diff
@@ -0,0 +1,48 @@
+use cookie::Cookie;
+use cookie::CookieJar;
+use cookie::ParseError;
+use futures::future;
+
+use crate::response::IntoResponse;
+use crate::{Extract, Request, Response, RouteMatch};
+
+#[derive(Clone, Debug)]
+pub struct Cookies {
+    content: CookieJar,
+}
+
+impl Cookies {
+    pub fn get(&self, name: &str) -> Option<&Cookie<'static>> {
+        self.content.get(name)
+    }
+}
+
+impl<S: 'static> Extract<S> for Cookies {
+    // Note: cannot use `existential type` here due to ICE
+    type Fut = future::Ready<Result<Self, Response>>;
+
+    fn extract(data: &mut S, req: &mut Request, params: &Option<RouteMatch<'_>>) -> Self::Fut {
+        match req.headers().get("Cookie") {
```

Comment:
I came up with an idea that might make the code easier to read:
1. Make `Result<CookieJar, ParseError>` at the `match req.headers().get("Cookie")` expression. For example, The `None` arm would return `Ok(CookieJar::new())`.
1. Transform that value into `Result<Cookies, Response>`.
1. Wrap it with `future::Ready`.

By doing this, we can easily see we're extracting `CookieJar` and then wrapping it with `Cookies`.

Reply:
good way. I updated the PR.